src/main/java/ru/timeconqueror/tcneiadditions/nei/TCNACrucibleRecipeHandler.java

```diff
@@ -109,27 +114,61 @@ public void drawExtras(int recipeIndex) {
 
         if (TCNAConfig.showResearchKey) {
             int y = 135;
-            String textToDraw = I18n.format("tcneiadditions.research.researchKey", recipe.researchKey);
-            for (Object text : Minecraft.getMinecraft().fontRenderer.listFormattedStringToWidth(textToDraw, 162)) {
+            String researchString = recipe.researchItem != null
+                    ? EnumChatFormatting.UNDERLINE
+                            + ResearchCategories.getCategoryName(recipe.researchItem.category) + " : "
+                            + recipe.researchItem.getName()
+                    : EnumChatFormatting.ITALIC + "null";
+            List listResearchString =
+                    Minecraft.getMinecraft().fontRenderer.listFormattedStringToWidth(researchString, 162);
+            this.ySize = listResearchString.size() * 11;
+            List<Object> list = new ArrayList<>();
+            list.add(StatCollector.translateToLocal("tcneiadditions.research.researchName") + ":");
+            list.addAll(listResearchString);
+            for (Object text : list) {
                 GuiDraw.drawStringC((String) text, 82, y, Color.BLACK.getRGB(), false);
                 y += 11;
             }
         }
     }
 
+    @Override
+    public List<String> handleTooltip(GuiRecipe gui, List<String> list, int recipeIndex) {
+        if (TCNAConfig.showResearchKey) {
+            if (GuiContainerManager.shouldShowTooltip(gui) && list.size() == 0) {
+                CrucibleCachedRecipe recipe = (CrucibleCachedRecipe) arecipes.get(recipeIndex);
+                Rectangle rectangle = getResearchRect(gui, recipeIndex);
+                Point mousePos = GuiDraw.getMousePosition();
+                if (rectangle.contains(mousePos.x, mousePos.y)) {
```

Comment:
nit: you can use `Rectangle#contains(Point)` instead of `#contains(int, int)`. Same for other handlers